draft-brunstrom-taps-impl.md

```diff
@@ -596,7 +596,7 @@ Connection lifetime for TCP translates fairly simply into the the abstraction pr
 
 If the application sends a Close, that can translate to a graceful termination of the TCP connection, which is performed by sending a FIN to the remote endpoint. If the application sends an Abort, then the TCP state can be closed abruptly, leading to a RST being sent to the peer.
 
-Without a layer of framing above TCP, the cleanest abstraction for sending and receiving Messages over TCP is to treat each direction of the stream of bytes as a single Message, terminated by a FIN. That means that if the application can make several Send calls to enqueue subsequent data chunks for the same Message to continue writing, but marking the Message complete corresponds to closing the sending stream. Similarly, when the application receives the final portion of a Message, it knows that the receiving stream has been closed.
+Without a layer of framing above TCP, the receiver side of the implementation has to rely on a supplied deframer to determine message boundaries. In the absence of such a deframer, a request by the application to only receive complete messages fails.
```

Comment:
A deframer *is* a layer of framing above TCP. So this is saying "if you don't have a deframer, use a deframer". I think that if you choose not to use a deframer, you should at the very least get an indication that a FIN has been received by the received content being marked as complete. That was actually a big problem for stream clients we had when we didn't give them this bit.

Reply:
Sorry about the language silliness, which I'll fix. But: a "layer of framing" could also be taken to mean a protocol - something that the sender and receiver both know about, e.g. HTTP/TCP inside the transport system.

The part I dislike about the original text is: "That means that if the application can make several Send calls to enqueue subsequent data chunks for the same Message to continue writing, but marking the Message complete corresponds to closing the sending stream."   Besides the sentence being broken, this again sounds like prohibiting my wish of "sending 50 short messages across a single connection".

I also generally find it strange to conflate "end of message" with "FIN" because the latter means "connection closed". You can always tell your applications that the remote endpoint has closed.

Reply:
There's a very important difference between "connection closed/finished" and one side sending a FIN. I can send a write-close after my initial request, and still keep the TCP connection up for a long time. If an application needs to know when the other send send the end-of-message/file, then that should be communicated in the Receive path, not in an event that "Connection Finished", which generally implies no more traffic at all.

Reply:
But we do not have a write-close so in the current API I guess the Close call is what will generate the FIN, correct? I am fine this being communicated in the receive path, but I think the "Connection Finished" will also follow from the FIN.